File: 2024/8/main.py
import sys
from collections import defaultdict
from itertools import combinations

data = open(sys.argv[1]).read().strip()
lines = data.split('\n')

grid = [list(line) for line in lines]


freqs = defaultdict(list)
for y, row in enumerate(grid):
    for x, cell in enumerate(row):
        if cell != '.':
            freqs[cell].append((y, x))


def in_bounds(y, x):
    return 0 <= y < len(grid) and 0 <= x < len(grid[0])


def get_obstacles(y1, x1, y2, x2, dy, dx):
    return y1-dy, x1-dx, y2+dy, x2+dx


def add_obstacles(seen, y1, x1, y2, x2):
    if in_bounds(y1, x1):
        seen.add((y1, x1))
    if in_bounds(y2, x2):
        seen.add((y2, x2))


seen1 = set()
seen2 = set()
for freq, coords in freqs.items():
    pairs = list(combinations(coords, 2))
    for (y1, x1), (y2, x2) in pairs:
        dy, dx = y2 - y1, x2 - x1
        oy1, ox1, oy2, ox2 = get_obstacles(y1, x1, y2, x2, dy, dx)
        add_obstacles(seen1, oy1, ox1, oy2, ox2)

        add_obstacles(seen2, y1, x1, y2, x2)
        dy, dx = y2 - y1, x2 - x1
        oy1, ox1, oy2, ox2 = get_obstacles(y1, x1, y2, x2, dy, dx)
        while in_bounds(oy1, ox1) or in_bounds(oy2, ox2):
            add_obstacles(seen2, oy1, ox1, oy2, ox2)
            oy1, ox1, oy2, ox2 = get_obstacles(oy1, ox1, oy2, ox2, dy, dx)

print(len(seen1))
print(len(seen2))

File: 2024/8/main.clj
(require '[clojure.string :as str])
(require '[clojure.math.combinatorics :as combo])

(def data (slurp (first *command-line-args*)))

(def lines (str/split data #"\n"))

(def grid (mapv (fn [line] (into [] line)) lines))

(def freqs (reduce (fn [freqs [y row]]
                     (reduce (fn [freqs [x cell]]
                               (if (not= cell \.)
                                 (update freqs cell (fnil conj []) [y x])
                                 freqs))
                             freqs
                             (map-indexed vector row)))
                   {}
                   (map-indexed vector grid)))

(defn in-bounds? [[y x]]
  (and (>= y 0) (< y (count grid))
       (>= x 0) (< x (count (grid 0)))))

(defn add-obstacles [seen [p1 p2]]
  (cond-> seen
    (in-bounds? p1) (conj p1)
    (in-bounds? p2) (conj p2)))

(defn get-obstacles [[y1 x1] [y2 x2] dy dx]
  [[(- y1 dy) (- x1 dx)] [(+ y2 dy) (+ x2 dx)]])

(def ans (->> freqs
              (reduce-kv (fn [seen _ coords]
                           (reduce (fn [seen [[y1 x1 :as p1] [y2 x2 :as p2]]]
                                     (add-obstacles seen
                                                    (get-obstacles p1 p2 (- y2 y1) (- x2 x1))))
                                   seen
                                   (combo/combinations coords 2)))
                         #{})
              count))

(def ans2 (->> freqs
               (reduce-kv (fn [seen _ coords]
                            (reduce (fn [seen [[y1 x1 :as p1] [y2 x2 :as p2]]]
                                      (let [dy (- y2 y1) dx (- x2 x1)]
                                        (loop [seen (conj seen p1 p2)
                                               [p1 p2] (get-obstacles p1 p2 dy dx)]
                                          (if (or (in-bounds? p1)
                                                  (in-bounds? p2))
                                            (recur (add-obstacles seen [p1 p2])
                                                   (get-obstacles p1 p2 dy dx))
                                            seen))))
                                    seen
                                    (combo/combinations coords 2)))
                          #{})
               count))

(println ans)
(println ans2)